dll/win32/shell32/dialogs/folder_options.cpp

```diff
@@ -1338,44 +1370,150 @@ InitializeFileTypesListCtrlColumns(HWND hDlgCtrl)
     HKEY hKey;
-    LONG nResult = RegOpenKeyExW(HKEY_CLASSES_ROOT, pszExt, 0, KEY_READ, &hKey);
-    if (nResult != ERROR_SUCCESS)
+    if (RegOpenKeyExW(HKEY_CLASSES_ROOT, pszExt, 0, KEY_READ, &hKey) != ERROR_SUCCESS)
         return FALSE;
 
-    WCHAR szValue[64];
+    // query "extfile" key name
+    WCHAR szValue[64] = { 0 };
     DWORD cbValue = sizeof(szValue);
-    nResult = RegQueryValueExW(hKey, NULL, NULL, NULL, LPBYTE(szValue), &cbValue);
+    RegQueryValueExW(hKey, NULL, NULL, NULL, LPBYTE(szValue), &cbValue);
     RegCloseKey(hKey);
-    if (nResult != ERROR_SUCCESS)
-        return FALSE;
 
+    // delete "extfile" key (if any)
     if (szValue[0])
         SHDeleteKeyW(HKEY_CLASSES_ROOT, szValue);
 
+    // delete ".ext" key
     return SHDeleteKeyW(HKEY_CLASSES_ROOT, pszExt) == ERROR_SUCCESS;
 }
 
-static
-VOID
-InsertFileType(HWND hDlgCtrl, WCHAR * szName, PINT iItem, WCHAR * szFile)
+static inline HICON
+DoExtractIcon(PFOLDER_FILE_TYPE_ENTRY Entry, LPCWSTR IconPath,
+              INT iIndex = 0, BOOL bSmall = FALSE)
+{
+    HICON hIcon = NULL;
+
+    if (iIndex < 0)
+    {
+        // A negative value will be interpreted as a negated resource ID.
+        iIndex = -iIndex;
+
+        INT cx, cy;
+        HINSTANCE hDLL = LoadLibraryExW(IconPath, NULL, LOAD_LIBRARY_AS_DATAFILE);
+        if (bSmall)
+        {
+            cx = GetSystemMetrics(SM_CXSMICON);
+            cy = GetSystemMetrics(SM_CYSMICON);
+        }
+        else
+        {
+            cx = GetSystemMetrics(SM_CXICON);
+            cy = GetSystemMetrics(SM_CYICON);
+        }
+        hIcon = HICON(LoadImageW(hDLL, MAKEINTRESOURCEW(iIndex), IMAGE_ICON,
+                                 cx, cy, 0));
+        FreeLibrary(hDLL);
+    }
+    else
+    {
+        // A positive value is icon index.
+        if (bSmall)
+            ExtractIconExW(IconPath, iIndex, NULL, &hIcon, 1);
+        else
+            ExtractIconExW(IconPath, iIndex, &hIcon, NULL, 1);
+    }
+    return hIcon;
+}
+
+static void
+DoFileTypeIconLocation(PFOLDER_FILE_TYPE_ENTRY Entry, LPCWSTR IconLocation)
+{
+    // Expand the REG_EXPAND_SZ string by environment variables
+    WCHAR szLocation[MAX_PATH + 32];
+    if (!ExpandEnvironmentStringsW(IconLocation, szLocation, _countof(szLocation)))
+    {
+        return;
+    }
+
+    INT nIndex = PathParseIconLocationW(szLocation);
+    Entry->hIconLarge = DoExtractIcon(Entry, szLocation, nIndex, FALSE);
+    Entry->hIconSmall = DoExtractIcon(Entry, szLocation, nIndex, TRUE);
+}
+
+static BOOL
+GetFileTypeIconsEx(PFOLDER_FILE_TYPE_ENTRY Entry, LPCWSTR IconLocation)
+{
+    Entry->hIconLarge = Entry->hIconSmall = NULL;
+
+    if (lstrcmpiW(Entry->FileExtension, L".exe") == 0 ||
+        lstrcmpiW(Entry->FileExtension, L".scr") == 0)
+    {
+        // It's an executable
+        HMODULE hDLL = shell32_hInstance;
+        const INT iIndex = IDI_SHELL_EXE;
+        Entry->hIconLarge = LoadIconW(hDLL, MAKEINTRESOURCEW(iIndex));
```

Comment:
Please put both the values of hDLL and iIndex inside the following two function calls: this will avoid the unnecessary creation of temporary variables.

Reply:
OK.